File: packages/game-engine/src/cards.ts
import type { CardCode } from "@parker/protocol";

export const SUITS = ["c", "d", "h", "s"] as const;
export const RANKS = ["2", "3", "4", "5", "6", "7", "8", "9", "T", "J", "Q", "K", "A"] as const;

export type Suit = (typeof SUITS)[number];
export type Rank = (typeof RANKS)[number];

export interface Card {
  code: CardCode;
  rank: Rank;
  suit: Suit;
  rankValue: number;
}

const rankValueMap = new Map<Rank, number>(RANKS.map((rank, index) => [rank, index + 2]));

export function parseCard(code: CardCode): Card {
  const [rank, suit] = code.split("") as [Rank, Suit];
  return {
    code,
    rank,
    suit,
    rankValue: rankValueMap.get(rank) ?? 0,
  };
}

export function createDeck(): Card[] {
  return SUITS.flatMap((suit) =>
    RANKS.map((rank) => parseCard(`${rank}${suit}` as CardCode)),
  );
}

File: internal/game/cards.go
package game

const (
	suitClubs    = "c"
	suitDiamonds = "d"
	suitHearts   = "h"
	suitSpades   = "s"
)

var suits = []string{suitClubs, suitDiamonds, suitHearts, suitSpades}
var ranks = []string{"2", "3", "4", "5", "6", "7", "8", "9", "T", "J", "Q", "K", "A"}

type Card struct {
	Code      CardCode
	Rank      string
	Suit      string
	RankValue int
}

var rankValueByRank = map[string]int{
	"2": 2,
	"3": 3,
	"4": 4,
	"5": 5,
	"6": 6,
	"7": 7,
	"8": 8,
	"9": 9,
	"T": 10,
	"J": 11,
	"Q": 12,
	"K": 13,
	"A": 14,
}

func ParseCard(code CardCode) Card {
	if len(code) != 2 {
		return Card{Code: code}
	}

	rank := string(code[0])
	suit := string(code[1])
	return Card{
		Code:      code,
		Rank:      rank,
		Suit:      suit,
		RankValue: rankValueByRank[rank],
	}
}

func CreateDeck() []Card {
	deck := make([]Card, 0, 52)
	for _, suit := range suits {
		for _, rank := range ranks {
			deck = append(deck, ParseCard(CardCode(rank+suit)))
		}
	}
	return deck
}
